From 787de32876a928d694c6249b429ced565d32209d Mon Sep 17 00:00:00 2001
From: Paul Sajna <hello@paulsajna.com>
Date: Sat, 22 Aug 2026 20:38:01 -0700
Subject: [PATCH] set targets

---
 CMakeLists.txt      | 1 +
 cmake/WineDLL.cmake | 1 +
 2 files changed, 2 insertions(+)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 1453500..27304c5 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -185,6 +185,7 @@ if(BUILD_WOW64_32)
         OUTPUT  "${PA32_SO}"
         COMMAND "${WINEGCC}" -shared
                 "${PA32_SPEC}"
+                --target=i386-windows
                 $<TARGET_OBJECTS:pipeasio32_unix_objs>
                 -lpthread -ldl ${PIPEWIRE_LINK_LIBRARIES} ${PIPEWIRE_LDFLAGS_OTHER}
                 ${_pa32_sanitize_libs}
diff --git a/cmake/WineDLL.cmake b/cmake/WineDLL.cmake
index 48d7435..aa58be5 100644
--- a/cmake/WineDLL.cmake
+++ b/cmake/WineDLL.cmake
@@ -213,6 +213,7 @@ function(add_wine_dll)
         OUTPUT  ${_so}
         COMMAND ${WINEGCC} -shared
                 ${WDL_SPEC}
+                --target=x86_64-windows
                 $<TARGET_OBJECTS:${_objlib}>
                 ${_winegcc_extra_flags}
                 ${_lflags}
-- 
2.55.0

